airflow/models/taskfail.py

```diff
@@ -37,6 +39,9 @@ class TaskFail(Base):
     end_date = Column(UtcDateTime)
     duration = Column(Integer)
 
+    task_instance = relationship("TaskInstance", lazy='joined', innerjoin=True)
```

Comment:
What would be the performance impact if we don’t set `lazy='joined'`? The TaskInstance object is large, so it’d be best if we could avoid pulling it in if possible.

Reply:
Thanks for the catch. I can see two modes 'joined' pulling object using join query to pull whole instance which as you have mentioned might be heavy and 'select' emitting query on attribute access. Querying TaskFail seems to be used in gantt and task_duration view. Only task_duration page accesses the execution_date attribute directly and gantt view converts object to dictionary that might trigger accessing execution_date.

https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html#relationship-loading-techniques

```
lazy loading - available via lazy='select' or the lazyload() option, this is the 
form of loading that emits a SELECT statement at attribute access time to lazily 
load a related reference on a single object at a time. Lazy loading is detailed at Lazy Loading

joined loading - available via lazy='joined' or the joinedload() option, this form of 
loading applies a JOIN to the given SELECT statement so that related rows are 
loaded in the same result set. Joined eager loading is detailed at Joined Eager Loading
```

```python
import sys

from airflow.settings import Session
from airflow.models import TaskFail

session = Session()
query = session.query(TaskFail)
task_fail = query.first()

print(f"Query : {query}")
print(task_fail)
print(sys.getsizeof(task_fail))

print(task_fail.execution_date)
print(sys.getsizeof(task_fail))
```

* joined

```
SELECT task_fail.id AS task_fail_id, task_fail.task_id AS task_fail_task_id, task_fail.dag_id AS task_fail_dag_id, task_fail.run_id AS task_fail_run_id, task_fail.map_index AS task_fail_map_index, task_fail.start_date AS task_fail_start_date, task_fail.end_date AS task_fail_end_date, task_fail.duration AS task_fail_duration, task_instance_1.try_number AS task_instance_1_try_number, dag_run_1.state AS dag_run_1_state, dag_run_1.id AS dag_run_1_id, dag_run_1.dag_id AS dag_run_1_dag_id, dag_run_1.queued_at AS dag_run_1_queued_at, dag_run_1.execution_date AS dag_run_1_execution_date, dag_run_1.start_date AS dag_run_1_start_date, dag_run_1.end_date AS dag_run_1_end_date, dag_run_1.run_id AS dag_run_1_run_id, dag_run_1.creating_job_id AS dag_run_1_creating_job_id, dag_run_1.external_trigger AS dag_run_1_external_trigger, dag_run_1.run_type AS dag_run_1_run_type, dag_run_1.conf AS dag_run_1_conf, dag_run_1.data_interval_start AS dag_run_1_data_interval_start, dag_run_1.data_interval_end AS dag_run_1_data_interval_end, dag_run_1.last_scheduling_decision AS dag_run_1_last_scheduling_decision, dag_run_1.dag_hash AS dag_run_1_dag_hash, dag_run_1.log_template_id AS dag_run_1_log_template_id, task_instance_1.task_id AS task_instance_1_task_id, task_instance_1.dag_id AS task_instance_1_dag_id, task_instance_1.run_id AS task_instance_1_run_id, task_instance_1.map_index AS task_instance_1_map_index, task_instance_1.start_date AS task_instance_1_start_date, task_instance_1.end_date AS task_instance_1_end_date, task_instance_1.duration AS task_instance_1_duration, task_instance_1.state AS task_instance_1_state, task_instance_1.max_tries AS task_instance_1_max_tries, task_instance_1.hostname AS task_instance_1_hostname, task_instance_1.unixname AS task_instance_1_unixname, task_instance_1.job_id AS task_instance_1_job_id, task_instance_1.pool AS task_instance_1_pool, task_instance_1.pool_slots AS task_instance_1_pool_slots, task_instance_1.queue AS task_instance_1_queue, task_instance_1.priority_weight AS task_instance_1_priority_weight, task_instance_1.operator AS task_instance_1_operator, task_instance_1.queued_dttm AS task_instance_1_queued_dttm, task_instance_1.queued_by_job_id AS task_instance_1_queued_by_job_id, task_instance_1.pid AS task_instance_1_pid, task_instance_1.executor_config AS task_instance_1_executor_config, task_instance_1.external_executor_id AS task_instance_1_external_executor_id, task_instance_1.trigger_id AS task_instance_1_trigger_id, task_instance_1.trigger_timeout AS task_instance_1_trigger_timeout, task_instance_1.next_method AS task_instance_1_next_method, task_instance_1.next_kwargs AS task_instance_1_next_kwargs 
FROM task_fail JOIN task_instance AS task_instance_1 ON task_instance_1.task_id = task_fail.task_id AND task_instance_1.dag_id = task_fail.dag_id AND task_instance_1.run_id = task_fail.run_id AND task_instance_1.map_index = task_fail.map_index JOIN dag_run AS dag_run_1 ON dag_run_1.dag_id = task_instance_1.dag_id AND dag_run_1.run_id = task_instance_1.run_id
```

* lazy='select'

```
SELECT task_fail.id AS task_fail_id, task_fail.task_id AS task_fail_task_id, task_fail.dag_id AS task_fail_dag_id, task_fail.run_id AS task_fail_run_id, task_fail.map_index AS task_fail_map_index, task_fail.start_date AS task_fail_start_date, task_fail.end_date AS task_fail_end_date, task_fail.duration AS task_fail_duration 
FROM task_fail
```

Reply:
The problem with `joined` is that it kicks in even if `execution_date` is never used (because a TaskFail cannot know beforehand if you’re going to use that attribute or not). If these are the two only usages, we should likely not use `joined`, but instead do an explicit join in those places.

Reply:
Agreed, IMO using joined with lazy is misleading and it's more of eager loading since join query is always made. There is 'select' that does select query on attribute access as per docs. In case of durations page we access and use execution_date. In case of gantt page the object is converted to dict but the execution_date attribute is not useful.  I am okay with using 'select' option or to have join query only in durations view as per original PR and remove the relationship. Thoughts?

Reply:
This disadvantage I see with 'select' is that it might endup in N+1 queries since execution_date is accessed inside for loop.

Reply:
Using `select` should be the way to go (or just omit `lazy` since IIRC `select` is the default). N+1 should be avoidable if we explicitly add a [`join`](https://docs.sqlalchemy.org/en/14/orm/query.html#sqlalchemy.orm.Query.join) on those queries.

Reply:
yes, select is the default. I am confused how join will help in avoiding N+1 since each attribute access inside for loop will trigger a query and if there are 10 objects it will trigger 10 select queries to task_instance table for execution_date. Can you please add the query/code?

Reply:
There are actually examples in the _Relationship Loading Technique_ page you linked 🙂 If we’re setting up `relationship` and `association_proxy`, the easiest solution is to tell SQLAlchemy to add the joins:

```python
session.query(TaskFail).options(joinedload(TaskFail.task_instance)).filter(...).all()
```

This would pre-load the `task_instance` relationship, and since `TaskInstance` already re-loads the `dag_run` relationship, it should make `TaskFail.execution_date` not generate additional SQL queries.

Reply:
Thanks, as I understand the default is to have 'select' by default for lazy in `TaskFail` and on durations page 'joinedload' can be used to load so we use joins in durations and efficiently leave to 'select' by default for `TaskFail` instances. I have updated the PR.